src/jni/org_netbsd_liblpm_LPM.c

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2016 Henry Rodrick <henry at holodisc org uk>
+ * All rights reserved.
+ *
+ * Use is subject to license terms, as specified in the LICENSE file.
+ */
+
+#include <inttypes.h>
+
+#include "org_netbsd_liblpm_LPM.h"
+#include "lpm.h"
+
+void dtor(void *, const void *, size_t, void *);
+
+void dtor(void *arg, const void *key, size_t len, void *val)
+{
+	JNIEnv *env = (JNIEnv*)arg;
+	(*env)->DeleteWeakGlobalRef(env, val);
+}
+
+JNIEXPORT jlong JNICALL Java_org_netbsd_liblpm_LPM_init
+  (JNIEnv *env, jobject obj)
+{
+	return (jlong) lpm_create();
+}
+
+JNIEXPORT void JNICALL Java_org_netbsd_liblpm_LPM_destroy
+  (JNIEnv *env, jobject obj, jlong lpm_ref)
+{
+	lpm_t* lpm = (lpm_t*)lpm_ref;
+	lpm_clear(lpm, dtor, env);
+	lpm_destroy(lpm);
+}
+
+JNIEXPORT void JNICALL Java_org_netbsd_liblpm_LPM_clear
+  (JNIEnv *env, jobject obj, jlong lpm_ref)
+{
+	lpm_clear((lpm_t*)lpm_ref, dtor, env);
+}
+
+
+JNIEXPORT jint JNICALL Java_org_netbsd_liblpm_LPM_insert
+  (JNIEnv *env, jobject obj, jlong lpm_ref, jstring cidr, jobject value)
+{
+	lpm_t *lpm;
+	jobject val_ref;
+	const char *cidr_s;
+	uint32_t addr[4];
+	size_t len;
+	unsigned pref;
+	int ret;
+
+	lpm = (lpm_t*)lpm_ref;
+
+	cidr_s = (*env)->GetStringUTFChars(env, cidr, NULL);
+	if (cidr_s == NULL) {
+		return -1;
+	}
+
+	ret = lpm_strtobin(cidr_s, addr, &len, &pref);
+	if (ret != 0) {
+		(*env)->ReleaseStringUTFChars(env, cidr, cidr_s);
+		return ret;
+
+	}
+
```

Comment:
You can perform `ReleaseStringUTFChars()` just after the `lpm_strtobin()` call and then simplify the remaining error branches.